### DISCUSSION TITLE
Discussion: New properties to add to module manifest?

### DISCUSSION THREAD

**Discussion by SteveL-MSFT** [PowerShell/PowerShell#15502] [Enhancements]
Is there a need for new module manifest properties?  Please indicate the proposed name of the property, description, and also examples of how it would be used?

This discussion came up during @PowerShell/powershell-committee review of https://github.com/PowerShell/PowerShell-RFC/pull/188 and we do not anticipate many new members, if any, being added to the module manifest.

**Comment by KirkMunro** [comment]
If a new member is proposed as part of an RFC, should it be listed here as well?

**Comment by ThomasNieto** [comment]
`CompatibleOS` from https://github.com/PowerShell/PowerShell/issues/9784 and https://github.com/PowerShell/PowerShell-RFC/pull/122

Max PS version support per https://github.com/PowerShell/PowerShell-RFC/pull/122. I expect it should be an enhancement of `PowerShellVersion` property to accept a hashtable like `RequiredModules`.

**Comment by SteveL-MSFT** [comment]
@KirkMunro yes, let's have a complete list here to discuss their merits

**Comment by KirkMunro** [comment]
From [PowerShell/PowerShell-RFC#220](https://github.com/PowerShell/PowerShell-RFC/pull/220):

|PropertyName|Type|Description|
|--|--|--|
|`OptionalFeatures`|`string[]`|Names of optional features that will be enabled within the scope of the script module|
|`DisabledOptionalFeatures`|`string[]`|Names of optional features that will be disabled within the scope of the script module|

The rationale behind these manifest entries is to allow optional features to be enabled or disabled by tool makers and module authors without impacting individual PowerShell users who want to use those tools. They allow module authors to guarantee their module behave a certain way regardless of what system they are run on, with one or more specific optional features enabled or disabled.

For more details why these are needed, read [PowerShell/PowerShell-RFC#220](https://github.com/PowerShell/PowerShell-RFC/pull/220).

For examples of possible optional features that module authors would care about, read [PowerShell/PowerShell-RFC#221](https://github.com/PowerShell/PowerShell-RFC/pull/221) and [PowerShell/PowerShell-RFC#187](https://github.com/PowerShell/PowerShell-RFC/pull/187).

**Comment by KirkMunro** [comment]
In addition to the above properties, it would be helpful if a manifest could be used without `psm1` files as much. For example, many modules include a `psm1` file whose sole purpose is to dot-source a bunch of `ps1` files.

To make that easier, module authors often just dot-source all `ps1` files in a folder (e.g. public or private). That could be dangerous though, because someone could drop a `ps1` file in a module folder resulting in additional code that is executed when a module is loaded. I've worked around that by only dot-sourcing files that have a signature that matches the signature on the `psm1` file, and reusing that code in a shared snippet in SnippetPx, but that's just creating more work to put things in place to make it easier to define modules.

It would be much better to identify the cases where a `psm1` file really shouldn't be necessary, and enhance `psd1` files so that they support those cases, making module authoring easier. That won't happen though, as long as we're stuck in legacy support mode.

**Comment by KirkMunro** [comment]
From [uservoice](https://windowsserver.uservoice.com/forums/301869-powershell/suggestions/11088036-add-requires-runasadministrator-ability-to-modul) and it just came up for me with a module I'm working on:

|PropertyName|Type|Description|
|--|--|--|
|`RunAsAdministrator`|`bool`|Identifies that the module can only be loaded in an elevated session. Only applies when run on Windows systems|

We have `#requires -RunAsAdministrator`, but that setting is the only `#requires` option that is not available in a manifest, so it should be made available there as well.

**Comment by vexx32** [comment]
I don't think there are a great many cases where loading a whole module _needs_ that, but I'm sure there are some. Having that info in the manifest would be a good idea, it makes it easier to determine the requirement before executing commands / importing the module.

**Comment by logicalextreme** [comment]
Just to pipe up here — I lost about half a day yesterday to investigating `#Requires -Assembly` having not been implemented, and only realised at the last moment before opening an issue that I'd written out that there were already issues for it which have all been closed (with the result that I can't tell whether the intention was to implement it, ignore it, or replace `#Requires` with module manifests entirely).

My feeling looking through the codebase for the first time was that it would make sense to have both `#Requires` and module manifests use the same validation (and loading, if applicable) methods for any shared functionality, as it seems like an obvious thing to abstract.

However my question is this — as I now haven't raised my issue for `#Requires -Assembly`, is anybody able to clarify that the intent is _not_ to implement it so that I can instead amend the documentation? I get the feeling it's not just me that's wasted time chasing a ghost that's been implemented as a parameter, documented and then not implemented.

Apologies for not raising a new issue but as I say there seem to be enough of them on the subject and they all pretty much lead to this one.
